LAMMPS-GUI Improvements

akohlmey opened the pull request:
**Summary**

Make behavior for selecting and configuring accelerators and threads in LAMMPS-GUI more
consistent. Improve efficiency of in run data processing. Increase the I/O buffer for capturing LAMMPS output from 1k to the maximum of 64k. Keep track of buffer usage and
if it is high, show dialog at the end of a run to either increase thermo or decrease the data polling interval.

Prevent crashes in several places and some minor redesigns.
Step version number to 1.7.0 in anticipation of the upcoming stable LAMMPS release.

As a bonus there is a new AI co-created banner logo. And I found and fixed a bunch of bugs including one bug causing a crash in a function imported from LAMMPS (so I had to copy the bugfix there, too)

![lammps-gui-with-banner](https://github.com/user-attachments/assets/39b3c21c-8288-4210-9c28-bbc3e5315ad2)


Also included are small additions to LAMMPS itself to be able to retrieve required information:
- new 'cpuuse' thermo keyword returning the %CPU used (like in LAMMPS performance summary output) so threading efficiency and impact be parasitic applications can be monitored
- expose the thermo interval and thermo_modify norm setting to `lammps_extract_setting()`

**Related Issue(s)**

N/A

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

The purpose of the changes are to make thread efficiency observable during a run, but also to be able to warn if LAMMPS-GUI has too much overhead or the simulation is running too fast and thermo output is too frequent to capture it reliably for plots.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the CMake based build system

akohlmey committed:
make some thermo settings accessible

akohlmey committed:
new thermo keyword cpuuse for %CPU utilization

akohlmey committed:
update handling of threads. make nthreads and LammpsGui class member

akohlmey committed:
only reset cached thermo data if first run or thermo style changed

this preserves last data from previous run for properties that are zero on the first step

akohlmey committed:
make handling of threads and accelerator selection more consistent, also buffer length

akohlmey committed:
cache spcpu and tpcpu values so these thermo keywords can be used multiple times

akohlmey committed:
update nthreads text field when accelerator is selected

the text is reset to 1 and editing disabled for None and Opt

akohlmey committed:
resolve hotkey conflict for selecting GPU accelerator package

akohlmey committed:
set LAMMPS-GUI version to 1.6.15

akohlmey committed:
support setting intel precision and gpu neigh and pair/only from preferences

akohlmey committed:
add a CPU usage percentage indicator to the status bar

akohlmey committed:
small improvements suggested by GitHub Copilot

akohlmey committed:
Merge remote-tracking branch 'github/develop' into lammps-gui-improvements

akohlmey committed:
update help index

akohlmey committed:
update LAMMPS-GUI documentation for most recent changes

akohlmey commented:
@simongravelle Here are some more changes to LAMMPS-GUI. Most are just internal to make its behavior more consistent and replace some ad hoc solutions with something that is more logical.

Also there are few new features:
- more settings via preferences for accelerators. E.g. to set precision for INTEL package or whether to run neighbor lists on host or the GPU for the GPU package
- a CPU utilization text field which can help to identify overloads and when LAMMPS is stalling.

I am still trying to figure out a way to determine the optimal polling interval for data. By extracting the thermo interval, steps per second (spcpu) thermo property, it should be possible to make some checks.

Update: I've found one more reason for slowdowns: the buffer for capturing the LAMMPS output was too small. I've increased it from 1k to 64k and simulations with small thermo interval run much more efficient. I'm keeping track of the buffer utilization and if it goes over one third, a warning dialog is shown suggestion to either increase thermo or decrease the data collection interval. The latter increases overhead due to more frequent locking/unlocking of the cached thermo data and thus also slows LAMMPS down a little bit, but not as much as the too small I/O buffer.

I've made new test executables and also updated the manual.

akohlmey committed:
reduce overhead and avoid stalling by increasing the pipe buffer from 1k to the maximum of 64k

akohlmey committed:
track buffer usage and add API to query the maximum buffer use ratio

akohlmey committed:
show warning dialog at end of run, if I/O buffer usage was very high

akohlmey committed:
should include C-math header since we use round()

akohlmey committed:
make precedence explicit

akohlmey committed:
add missing screenshot

akohlmey committed:
implement suggestion from GitHub CoPilot to make code consistent and simpler

akohlmey committed:
remove dead code and apply changes for clean compilation w/o OpenMP

akohlmey committed:
display new LAMMPS-GUI banner image in docs and empty editor windows

akohlmey committed:
fix typo

akohlmey committed:
avoid crash from out-of-range access to string for partially quoted text

akohlmey committed:
must clear variables before starting a new run since "clear" does not delete them

akohlmey committed:
update changelog

akohlmey committed:
simplify

akohlmey committed:
small tweaks to error message boxes

akohlmey committed:
save and restore PATH and LD_LIBRARY_PATH, so that xdg-open can launch a native executable

akohlmey committed:
fix typo

akohlmey committed:
document Relaunch LAMMPS Instance

akohlmey committed:
prevent using Restart LAMMPS while running and rename to Relaunch LAMMPS

akohlmey committed:
delete app bundle folder. It will be recreated by macdeplotqt.

akohlmey committed:
must use a platform specific copy command

simongravelle commented:
Hi @akohlmey 

Thank you for all these updates to the GUI.

> I've made new test executables and also updated the manual.

I have been playing with the new version of the GUI. I did not encounter any bog or any unexpected behavior. I like the new logo that appear from the start and disappear as soon as one clicks.

I don't have much to say concerning the performance, as I am really no expert of accelerator, although as a user I appreciate that irrelevant options now appear as "gray".

Simon

akohlmey commented:
> Thank you for all these updates to the GUI.

Still looking into issues that were reported. Some are really strange like the Windows version was stalling when trying to download tutorial files on Windows due to a bug in the Fedora 41/42 OpenSSL cross-compiler package. Now I have to use a Fedora 40 container and rebuild all recent Windows packages so that the LAMMPS geturl command will work properly.

Now I'm working on a workaround for the workaround for handling symbolic links in the git repo on Windows.

akohlmey committed:
there is not much meaning in completing capture in the destructor.

the user should have already called EndCapture() if needed.
calling it in the destructor just causes memory corruption.

akohlmey committed:
automatically quit the entire application when closing the editor window

avoid crashes and segfaults when simulation is still running

akohlmey committed:
must use recursive delete on app bundle folder

akohlmey committed:
redesign the General settings preference tab and add defaults for tutorial buttons

akohlmey committed:
to update the time stamp on the app bundle purge it during configure

akohlmey committed:
set LAMMPS-GUI version number to 1.7

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
 ----
 
 LAMMPS-GUI tries to provide an experience similar to what people
```
```suggestion
LAMMPS-GUI aims to provide the traditional experience of running LAMMPS: 
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
 LAMMPS-GUI tries to provide an experience similar to what people
-traditionally would have running LAMMPS using a command-line window and
-the console LAMMPS executable but just rolled into a single executable:
+traditionally would have when running LAMMPS using a text editor, the
```
```suggestion
using a text editor, a
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
-traditionally would have running LAMMPS using a command-line window and
-the console LAMMPS executable but just rolled into a single executable:
+traditionally would have when running LAMMPS using a text editor, the
+command-line window, and launching the LAMMPS executable but just rolled
```
```suggestion
command-line window, and launching the LAMMPS executable, but integrated
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
-the console LAMMPS executable but just rolled into a single executable:
+traditionally would have when running LAMMPS using a text editor, the
+command-line window, and launching the LAMMPS executable but just rolled
+into a single executable:
```
May be a single "application" instead of a single "executable" ?

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
+command-line window, and launching the LAMMPS executable but just rolled
+into a single executable:
 
 - writing & editing LAMMPS input files with a text editor
```
```suggestion
- Write and edit LAMMPS input files using a built-in text editor.
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
+into a single executable:
 
 - writing & editing LAMMPS input files with a text editor
 - run LAMMPS on those input file with selected command-line flags
```
```suggestion
- Run LAMMPS on these input files with selectable command-line flags.
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
 - writing & editing LAMMPS input files with a text editor
 - run LAMMPS on those input file with selected command-line flags
-- extract data from the created files and visualize it with and
+- extract data from the created files and visualize it with
```
Extract data from the generated files for visualization using external software.

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+be selected and then moved and dropped on the LAMMPS-GUI executable, and
+LAMMPS-GUI will launch and read the file into its buffer.  In many cases
+LAMMPS-GUI will be integrated into the graphical desktop environment and
+can be launched like other applications.
```
```suggestion
can be launched just like any other applications.
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-graphical desktop environment and can be launched like other
-applications.
+clicking on such files using a graphical file manager.  LAMMPS-GUI also
+has support for drag-n-drop for opening inputs, i.e.  an input file can
```
```suggestion
supports drag-and-drop: an input file can
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-applications.
+clicking on such files using a graphical file manager.  LAMMPS-GUI also
+has support for drag-n-drop for opening inputs, i.e.  an input file can
+be selected and then moved and dropped on the LAMMPS-GUI executable, and
```
```suggestion
be dragged onto the LAMMPS-GUI executable, which will then
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+clicking on such files using a graphical file manager.  LAMMPS-GUI also
+has support for drag-n-drop for opening inputs, i.e.  an input file can
+be selected and then moved and dropped on the LAMMPS-GUI executable, and
+LAMMPS-GUI will launch and read the file into its buffer.  In many cases
```
```suggestion
launch and load the file into its buffer.  In many cases,
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -22,15 +26,16 @@ This document describes **LAMMPS-GUI version 1.6**.
 
-That procedure is quite effective for people proficient in using the
+That traditional procedure is effective for people proficient in using the
 command-line, as that allows them to use tools for the individual steps
```
```suggestion
command-line, as it allows them to use tools for the individual steps
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-simple LAMMPS simulations.  It is very suitable for tutorials on LAMMPS
-since you only need to learn how to use a single program for most tasks
-and thus time can be saved and people can focus on learning LAMMPS.
+LAMMPS simulations since you can start with learning LAMMPS right away.
```
```suggestion
LAMMPS simulations, as they can focus on learning LAMMPS right away.
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-since you only need to learn how to use a single program for most tasks
-and thus time can be saved and people can focus on learning LAMMPS.
+LAMMPS simulations since you can start with learning LAMMPS right away.
+It is therefore very suitable for tutorials on LAMMPS, since you only
```
```suggestion
It is particularly well-suited for LAMMPS tutorials, since users only
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-and thus time can be saved and people can focus on learning LAMMPS.
+LAMMPS simulations since you can start with learning LAMMPS right away.
+It is therefore very suitable for tutorials on LAMMPS, since you only
+need to learn how to use a single program that is available as
```
```suggestion
need to work with a single, ready-to-use program,
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+LAMMPS simulations since you can start with learning LAMMPS right away.
+It is therefore very suitable for tutorials on LAMMPS, since you only
+need to learn how to use a single program that is available as
+ready-to-use, pre-compiled packages for popular operating systems
```
```suggestion
available as pre-compiled packages for popular operating systems
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+need to learn how to use a single program that is available as
+ready-to-use, pre-compiled packages for popular operating systems
+(Linux, macOS, Windows), for most of the tasks.  This saves time and
+people can focus on learning LAMMPS instead of having to compile LAMMPS,
```
```suggestion
allows users to focus on learning LAMMPS itself, without needing to compile it first,
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+ready-to-use, pre-compiled packages for popular operating systems
+(Linux, macOS, Windows), for most of the tasks.  This saves time and
+people can focus on learning LAMMPS instead of having to compile LAMMPS,
+learn how to use the command line and a text editor.
```
```suggestion
learn how to use the command line, or learn to use a separate text editor.
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
-the functionality of LAMMPS-GUI with external tools.  That said, LAMMPS-GUI
-has some unique functionality that is not found elsewhere:
+Yet the basic control flow remains similar to running LAMMPS from the
+command line and thus the barrier for replacing part of the
```
```suggestion
command line, so the barrier to replacing parts of the
```

simongravelle commented on doc/src/Howto_lammps_gui.rst:
```diff
@@ -42,35 +47,44 @@ window or using external programs, let alone writing scripts to extract
+Yet the basic control flow remains similar to running LAMMPS from the
+command line and thus the barrier for replacing part of the
+functionality of LAMMPS-GUI with external tools is low.  That said,
+LAMMPS-GUI has some unique functionality that is not easily found
```
```suggestion
LAMMPS-GUI offers some unique features that are not easily found
```

simongravelle reviewed (commented):
@akohlmey, I realize that my comments are too numerous while also being minor, so it must be really annoying for you. 

Would it be OK/best for you if I just commit my change directly? At least for the minor one.

akohlmey commented:
> @akohlmey, I realize that my comments are too numerous while also being minor, so it must be really annoying for you.
> 
> Would it be OK/best for you if I just commit my change directly? At least for the minor one.

The problem is that you have overlapping edits, so it is difficult to follow and the GitHub web interface does not really support this, since it always compares to the original.

I don't agree with all your changes or would want to make additional modifications. The simplest way would be that you edit the source file and attach it here and I can then do a selective merge and edit and commit the combined changes. The more formal approach would be to submit a pull request for this specific branch in https://github.com/akohlmey/lammps/

akohlmey committed:
apply and edit some changes suggested by @simongravelle

akohlmey committed:
remove rubberband feature since undoing the zoom doesn't work properly

akohlmey committed:
refactor chart viewer title line. Now has two rows. display units and thermo norm

akohlmey committed:
add rangeslider custom widget

akohlmey committed:
add stop button to chart display

akohlmey committed:
Revert "add stop button to chart display"

This reverts commit 1bb746aa877edace69881cdab821739886d9270e.

akohlmey committed:
add rangeslider to chart window to plot subsets of data

akohlmey committed:
Merge remote-tracking branch 'github/develop' into lammps-gui-improvements

akohlmey committed:
hack trying to appease macOS

akohlmey committed:
intergrate the rest of simon's suggestions

akohlmey committed:
spelling a re-wrap

akohlmey committed:
add tooltips

akohlmey committed:
avoid crashes when switching to a new file or a new empty document and close all windows

akohlmey committed:
fix typo

akohlmey committed:
avoid uninitialized access to normflag

akohlmey committed:
add a couple horizontal separator lines

akohlmey committed:
update docs and screen shot

akohlmey committed:
minor doc tweaks for readability

akohlmey commented:
@simongravelle I think I have now integrated all your suggestions for the LAMMPS-GUI Howto.
There also is another round of bugfixes and avoiding crashes under unusual circumstances (like switching to a new document while LAMMPS is still running). Also log and chart windows etc. should be closed. I spent considerable amounts of time trying to make the resetZoom button work (again) after using the so-called rubberband zoom, but ultimately failed and implemented a different approach using two "double edged sliders". Also that took some time. I found sources for two suitable custom widgets that implement this on the web, but the first was not good enough and I had to give up on it after a while. The second needed a little work, too, but I could make it work the way I was looking for (plus it had an unexpected "sliding window" feature when to drag the middle of the slider around).

With all those changes, we should update the tutorial to require LAMMPS-GUI version 1.7.
Do we also need to update any screenshots?

simongravelle commented:
> With all those changes, we should update the tutorial to require LAMMPS-GUI version 1.7.
> Do we also need to update any screenshots?

OK, I am currently applying the reviewer comments, so I can update the tutorial to the version 1.7, no problem (unless may be it would be wiser to wait also for the stable release, and make all tutorials compatible with both stable release of LAMMPS and GUI 1.7?).

If there is a need for updated screenshots, I will do it as well.

akohlmey commented:
> and make all tutorials compatible with both stable release of LAMMPS and GUI 1.7?).

As far as I can tell, there is nothing in the tutorial that would not work with the current and the future stable version. The most critical is probable tutorial 8.
When I am tired of stomping on bugs, I will do a walkthrough of the entire set.

akohlmey commented:
@simongravelle FWIW, I have been backporting all LAMMPS-GUI changes the stable release see PR #4629. If desired, I can make binaries for that version as well. I usually only compile that one locally, since use the two versions to also check for differences between Qt 5.x and 6.x.
Although the biggest challenge is usually Ubuntu 20.04LTS since it ships with Qt 5.12 instead of the last stable 5.x version 5.15.

akohlmey committed:
fix halt may only reset the timer timeout, if it trigged the timeout itself

akohlmey committed:
enable option to open the tutorial webpage for tutorial 8

akohlmey committed:
add ticks to range sliders

akohlmey committed:
cosmetic suggestions from GitHub Copilot

akohlmey committed:
fully restore original script code

akohlmey committed:
only restore old paths, if they exist, i.e. we are called from within LAMMPS-GUI

akohlmey committed:
add small tweaks to the LAMMPS-GUI banner and how it is shown

akohlmey committed:
silence compiler warnings and apply small corrections

akohlmey committed:
Merge remote-tracking branch 'github/develop' into lammps-gui-improvements

megmcca reviewed (approved):
approved!
